src/Http/Routing/src/Matching/CandidateSet.cs

```diff
@@ -302,39 +167,197 @@ public void ReplaceEndpoint(int index, Endpoint endpoint, RouteValueDictionary v
+        {
+            // Friendliness for inlining
+            if ((uint)index >= Count)
+            {
+                ThrowIndexArgumentOutOfRangeException();
+            }
+
+            if (endpoints == null)
+            {
+                ThrowArgumentNullException(nameof(endpoints));
+            }
+
+            if (comparer == null)
+            {
+                ThrowArgumentNullException(nameof(comparer));
+            }
+
+            // First we need to verify that the score of what we're replacing is unique.
+            ValidateUniqueScore(index);
+
+            switch (endpoints.Count)
             {
                 case 0:
-                    _state0 = new CandidateState(endpoint, values, _state0.Score);
+                    ReplaceEndpoint(index, null, null);
                     break;
 
                 case 1:
-                    _state1 = new CandidateState(endpoint, values, _state1.Score);
+                    ReplaceEndpoint(index, endpoints[0], _candidates[index].Values);
                     break;
 
-                case 2:
-                    _state2 = new CandidateState(endpoint, values, _state2.Score);
-                    break;
+                default:
 
-                case 3:
-                    _state3 = new CandidateState(endpoint, values, _state3.Score);
-                    break;
+                    var score = GetOriginalScore(index);
+                    var values = _candidates[index].Values;
+
+                    // Adding candidates requires expanding the array and computing new score values for the new candidates.
+                    var original = _candidates;
+                    var candidates = new CandidateState[original.Length - 1 + endpoints.Count];
+                    _candidates = candidates;
+
+                    // Since the new endpoints have an unknown ordering relationship to each other, we need to:
+                    // - order them
+                    // - assign scores
+                    // - offset everything that comes after
+                    //
+                    // If the inputs look like:
+                    //
+                    // score 0: A1
+                    // score 0: A2
+                    // score 1: B
+                    // score 2: C <-- being expanded
+                    // score 3: D
+                    //
+                    // Then the result should look like:
+                    //
+                    // score 0: A1
+                    // score 0: A2
+                    // score 1: B
+                    // score 2: `C1
+                    // score 3: `C2
+                    // score 4: D
+
+                    // Candidates before index can be copied unchanged.
+                    for (var i = 0; i < index; i++)
+                    {
+                        candidates[i] = original[i];
+                    }
+
+                    var buffer = endpoints.ToArray();
```

Comment:
This would happen per-request right? I worry about CMSes out there that have tens of thousands of endpoints. Large blogs, news sites, etc. Is it worth having an API that allows a user to pass in a pre-sorted list of endpoints that they're cached?

Reply:
I don't think passing in 1000 endpoints here is viable whether we pre-sort them or not. You're asking us to iterate 1000 endpoints multiple times and process policies on all of them. At that point you're holding it wrong. 

I don't think this is a implementation strategy we should try to enable. If you're writing a CMS with 1000s of endpoints you need to apply some kind of lookup yourself and not *just* use `IEndpointSelectorPolicy`. I haven't heard anyone ask for this.